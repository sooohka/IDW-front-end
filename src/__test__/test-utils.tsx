/* eslint-disable @typescript-eslint/no-shadow */
/* eslint-disable no-param-reassign */
import { render, RenderOptions } from "@testing-library/react";
import React, { FC, ReactElement } from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import renderer from "react-test-renderer";
import store from "../store";
import GlobalStyle from "../style/GlobalStyle";
import theme from "../style/theme";

const AllTheProviders: FC = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      <Provider store={store}>
        <BrowserRouter>
          <GlobalStyle />
          {children}
        </BrowserRouter>
      </Provider>
    </ThemeProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const thunk =
  ({ dispatch, getState }: any) =>
  (next: any) =>
  (action: any) => {
    if (typeof action === "function")
      // 여기서 반환되는 action은 thunk함수임
      return action(dispatch, getState);
    return next(action);
  };

const useStore = (stateName: string, initState: any, reducer: any) => {
  const store = {
    dispatch: jest.fn((action: any) => {
      initState = reducer(initState, action);
    }),
    getState: jest.fn(() => ({ [stateName]: initState })),
  };
  const next = jest.fn();
  const makeThunk = (action: any) => thunk(store)(next)(action);
  return { store, next, makeThunk };
};

const createWithThemeProvider = (EL: ReactElement) => {
  return renderer.create(
    <ThemeProvider theme={theme}>
      <GlobalStyle />
      {EL}
    </ThemeProvider>,
  );
};

const redux = { thunk, useStore };
export { redux, customRender as render, createWithThemeProvider };

export * from "@testing-library/react";
