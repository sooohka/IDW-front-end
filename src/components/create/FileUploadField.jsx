import axios from "axios";
import { useField } from "formik";
import PropTypes from "prop-types";
import React, { useCallback, useEffect, useState } from "react";
import { useDropzone } from "react-dropzone";
import uuid from "uuid";
import styled, { css } from "styled-components";
import theme from "../../style/theme";
import useMount from "../../utils/hooks/useMount";
import { readAsDataUrl } from "../../utils/lib/file";
import { ReactComponent as UploadIcon } from "../../assets/icons/cloud-upload-alt-solid.svg";
import { ReactComponent as Sort } from "../../assets/icons/sort-up-solid.svg";
import FileUploadWithProgress from "./FileUploadWithProgress";

const Container = styled.div`
  min-height: 150%;
`;

const DropZone = styled.div`
  border: 3px dashed;
  border-color: ${() => theme.colors.primary};
  height: 15rem;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  flex-direction: column;
  & > * {
    transform: ${({ isAccepting }) => isAccepting && "scale(1.5)"};
    opacity: ${({ isAccepting }) => isAccepting && 0.3};
  }
  & > p {
    font-weight: bold;
    color: ${() => theme.colors.primary};
  }
`;

const FileList = styled.div`
  margin: 1rem 0 0 0;
  border: 2px solid black;
  border-radius: 5px;
  min-height: 3rem;
  position: relative;
`;

const FileListTitle = styled.h3`
  flex: 1;
  padding-left: 1rem;
`;

const FileListHeader = styled.div`
  display: flex;
  align-items: center;
  cursor: pointer;
`;

const Files = styled.div`
  transition: all;
  transform-origin: top;
  ${({ isFolded }) => {
    if (isFolded)
      return css`
        transform: scaleY(0);
        max-height: 0;
      `;
    return css`
      transform: scaleY(1);
      padding: 2rem 0 0;
      max-height: 10000px;
    `;
  }}
`;
const IconWrapper = styled.div`
  cursor: pointer;
  & > svg {
    transition: all 0.2s;
    ${({ isFolded }) => {
      if (isFolded) {
        return css`
          transform: rotate(180deg);
        `;
      }
      return css`
        transform: rotate(0deg) translateY(30%);
      `;
    }}
  }
`;
const FileUploadField = ({ formikName, setIsFileUploading, buttonEl }) => {
  const [, , helpers] = useField(formikName);
  const [files, setFiles] = useState([]);
  // const [submittedFiles, setSubmittedFiles] = useState([]); // { name: "", url: "" } =>helper.setValue Ìï¥ÏïºÌï®
  const [isAccepting, setIsAccepting] = useState(false);
  const { isMount } = useMount();
  const [isFolded, setIsFolded] = useState(true);

  const handleUpload = useCallback(async (_file, setProgress) => {
    const {
      file: { lastModified, lastModifiedDate, name, path, size, type },
      id: currentFileId,
      url: unused,
      isSubmitted,
      error: { status, message: m },
    } = _file;
    const copiedFile = { name, lastModified, lastModifiedDate, path, size, type };

    try {
      const fileDataUrl = await readAsDataUrl(_file.file);
      const request = { file: { name, contentType: type, dataUri: fileDataUrl } };

      if (process.env.REACT_APP_ENV !== "local") {
        alert("env=localÏûÖÎãàÎã§.");
        console.log("request:", request);
        return { hasError: false, message: "env=local", file: copiedFile, url: "" };
      }

      const response = await axios.post("https://dogemdas2c.execute-api.ap-northeast-2.amazonaws.com/v1", request, {
        onUploadProgress: (prog) => setProgress(Math.round(prog.loaded * 100) / prog.total),
      });

      // TODO:ÌååÏùº ÏïàÎ≥¥ÎÇ¥ÏßàÎïå ÏóêÎü¨Ï≤òÎ¶¨
      console.log(response);

      const {
        message,
        result: { url, ContentType, bucketUrl, locations },
      } = response.data;

      const setFile = (v) => {
        if (v.id === currentFileId) return { ...v, isSubmitted: true, url, fullUrl: `${bucketUrl}${locations.small}` };
        return { ...v };
      };

      setFiles((prev) => prev.map(setFile));
      return {
        hasError: false,
        message: message || "submitted",
        file: { name, lastModified, lastModifiedDate, path, size, type },
        url: `${bucketUrl}${locations.small}`,
      };
    } catch (err) {
      console.error(err.message);
      // TODO: fileUploadWithProgress error handling
      const message = err.response?.data?.error?.message || err.message || "something went wrongüòÖ ";
      const setFile = (v) => {
        if (v.id === currentFileId) return { ...v, isSubmitted: true, error: { status: true, message } };
        return { ...v };
      };
      setFiles((prev) => prev.map(setFile));
      return { hasError: true, message, file: { name, lastModified, lastModifiedDate, path, size, type } };
    }
  }, []);

  useEffect(() => {
    if (buttonEl.current) buttonEl.current.scrollIntoView({ behavior: "smooth", block: "end" });
  }, [buttonEl, isFolded]);

  useEffect(() => {
    setIsFileUploading(true);
  }, [files, setIsFileUploading]);

  useEffect(() => {
    if (isMount) return;
    console.log(`%c state submittedFiles in FileUploadField changed`, "background-color:pink;font-size:15px;font-weight:bold;color:black");

    const isEveryFileSubmitted = files.every((v) => v.isSubmitted);
    if (isEveryFileSubmitted) setIsFileUploading(false);
    else helpers.setError("ÌååÏùºÏù¥ ÏóÖÎ°úÎìúÏ§ëÏûÖÎãàÎã§.");

    const formattedFormikFile = files.map((v) => ({ name: v.name, url: v.url }));
    helpers.setValue(formattedFormikFile, true);
  }, [files, isMount, setIsFileUploading]);

  const onDrop = useCallback((acceptedFiles, rejectedFiles) => {
    // TODO: ÏóêÎü¨Î©îÏãúÏßÄ Í∞úÏÑ†
    if (rejectedFiles.length > 0) {
      console.log(rejectedFiles);
      const rejected = rejectedFiles.map((v) => ({ name: v.file.name, errors: v.errors[0].message }));
      alert(JSON.stringify({ rejectedFiles: rejected }, null, 2));
    }
    const formedAcceptedFiles = acceptedFiles.map((file) => ({
      file,
      isSubmitted: false,
      id: uuid(),
      url: "",
      fullUrl: "",
      error: { status: false, message: "" },
    }));
    setFiles((prev) => [...prev, ...formedAcceptedFiles]);
    setIsAccepting(false);
  }, []);

  const onDragEnter = () => setIsAccepting(true);
  const onDragLeave = () => setIsAccepting(false);

  const handleValidation = (_file) => {
    if (files.find(({ file }) => file.name === _file.name && file.lastModified === _file.lastModified)) return { message: "Í∞ôÏùÄ ÌååÏùºÏùÄ ÏóÖÎ°úÎìúÎ∂àÍ∞ÄÎä•Ìï©ÎãàÎã§." };
    if (_file.size >= 5000000) return { message: "ÌååÏùºÏùÄ ÏµúÎåÄ 4.9MBÍπåÏßÄ ÏóÖÎ°úÎìú Í∞ÄÎä•Ìï©ÎãàÎã§." };
    if (files.length > 50) return { message: "ÌååÏùºÏùÄ ÏµúÎåÄ 50Ïû• ÏóÖÎ°úÎìú Í∞ÄÎä•Ìï©ÎãàÎã§." };

    return null;
  };
  const handleDelete = useCallback(
    (id) => () => {
      setFiles((prev) => prev.filter((file) => file.id !== id));
    },
    [],
  );

  const { getRootProps, getInputProps } = useDropzone({
    accept: "image/*",
    maxFiles: 50,
    onDrop,
    onDragEnter,
    onDragLeave,
    validator: handleValidation,
  });

  return (
    // <Template
    //   isFolded={isFolded}
    //   setIsFolded={setIsFolded}
    //   handleDelete={handleDelete}
    //   handleUpload={handleUpload}
    //   files={React.useMemo(() => files, [files])}
    //   isAccepting={isAccepting}
    //   getRootProps={getRootProps}
    //   getInputProps={getInputProps}
    // />
    <Container>
      <DropZone isAccepting={isAccepting} {...getRootProps()}>
        <input {...getInputProps()} />
        <UploadIcon fill={theme.colors.primary} width={50} height={50} />
        <p>Drag and Drop or click here to upload</p>
      </DropZone>

      {/* ListFiles */}
      <FileList>
        <FileListHeader onClick={() => setIsFolded((prev) => !prev)}>
          <FileListTitle>ÌååÏùºÎì§</FileListTitle>
          <IconWrapper isFolded={isFolded}>
            <Sort width={30} height={30} />
          </IconWrapper>
        </FileListHeader>
        <Files isFolded={isFolded}>
          {files.map((file) => (
            <FileUploadWithProgress handleUpload={handleUpload} handleDelete={handleDelete} file={file} key={file.id} />
          ))}
          {files.length > 5 && (
            <div style={{ display: "flex", justifyContent: "flex-end" }}>
              <IconWrapper onClick={() => setIsFolded(true)} isFolded={isFolded}>
                go Top
                <Sort width={30} height={30} />
              </IconWrapper>
            </div>
          )}
        </Files>
      </FileList>
    </Container>
  );
};
FileUploadField.propTypes = {
  buttonEl: PropTypes.shape({ current: PropTypes.object }).isRequired,
  setIsFileUploading: PropTypes.func.isRequired,
  formikName: PropTypes.string.isRequired,
};
FileUploadField.defaultProps = {};

export default FileUploadField;
