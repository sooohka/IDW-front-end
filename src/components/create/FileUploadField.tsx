import { useFormikContext } from "formik";
import React, { useEffect, useState } from "react";
import { FileRejection, useDropzone } from "react-dropzone";
import styled from "styled-components";
import * as uuid from "uuid";
import { ReactComponent as UploadIcon } from "../../assets/icons/cloud-upload-alt-solid.svg";
import globalTheme from "../../style/theme";
import FileList from "./FileList";
import NewFileUploadWithProgress from "./FileUploadWithProgress";

const Container = styled.div`
  min-height: 150%;
`;

const DropZone = styled.div<{ isAccepting: boolean }>`
  border: 3px dashed;
  border-color: ${({ theme }) => theme.colors.primary};
  height: 15rem;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  flex-direction: column;
  & > * {
    transform: ${({ isAccepting }) => isAccepting && "scale(1.5)"};
    opacity: ${({ isAccepting }) => isAccepting && 0.3};
  }
  & > p {
    font-weight: bold;
    color: ${({ theme }) => theme.colors.primary};
  }
`;

interface IProps {
  formikName: string;
  setIsFileUploading: React.Dispatch<React.SetStateAction<boolean>>;
}

const NewFileUploadField: React.FC<IProps> = ({ formikName, setIsFileUploading }) => {
  const { setFieldValue } = useFormikContext();
  const [imageFiles, setImageFiles] = useState<TargetFile[]>([]);

  /* drag and drop필드에 접근했을때 */
  const [isAccepting, setIsAccepting] = useState(false);

  const handleDelete = (id: string) => () => {
    setImageFiles((prev) => prev.filter((file) => file.id !== id));
  };

  const onDragEnter = () => setIsAccepting(true);

  const onDragLeave = () => setIsAccepting(false);

  const onDrop = (acceptedFiles: File[], rejectedFiles: FileRejection[]) => {
    const accFiles = acceptedFiles.map((acceptedFile) => ({
      file: acceptedFile,
      errors: [],
      id: uuid.v4(),
      isSubmitted: false,
      url: "",
    }));

    const rejFiles = rejectedFiles.map((rejectedFile) => ({
      ...rejectedFile,
      id: uuid.v4(),
      isSubmitted: false,
      url: "",
    }));

    setImageFiles((prevFiles) => [
      ...prevFiles.filter((prevFile) => prevFile.errors.length === 0),
      ...accFiles,
      ...prevFiles.filter((prevFile) => prevFile.errors.length > 0),
      ...rejFiles,
    ]);
    setIsAccepting(false);
  };

  useEffect(() => {
    if (imageFiles.every((imageFile) => imageFile.isSubmitted && !imageFile.errors.length)) {
      const formedFiles = imageFiles.map((imageFile) => ({
        name: imageFile.file.name,
        url: imageFile.url,
      }));
      setFieldValue(formikName, formedFiles);
      setIsFileUploading(false);
    } else {
      setIsFileUploading(true);
    }
  }, [formikName, imageFiles, setFieldValue, setIsFileUploading]);

  const handleValidation = (file: File) => {
    if (imageFiles.find((imageFile) => imageFile.file.name === file.name)) {
      return {
        code: "duplicate-file",
        message: "중복된 파일은 허용이 안됩니다.",
      };
    }
    if (file.size / 1024 / 1024 > 5)
      return {
        code: "file-too-large",
        message: `파일은 5mb를 넘을 수 없습니다`,
      };
    return null;
  };

  const { getRootProps, getInputProps } = useDropzone({
    accept: "image/*",
    maxFiles: 50,
    onDrop,
    onDragEnter,
    onDragLeave,
    validator: handleValidation,
  });

  return (
    <Container>
      <DropZone {...getRootProps()} isAccepting={isAccepting}>
        <input {...getInputProps()} />
        <UploadIcon fill={globalTheme.colors.primary} width={50} height={50} />
        <p>Drag and Drop or click here to upload</p>
      </DropZone>

      {/* ListFiles */}

      <FileList>
        {imageFiles.map((imageFile) => (
          <NewFileUploadWithProgress
            imageFile={imageFile}
            key={imageFile.id}
            handleDelete={handleDelete}
            setImageFiles={setImageFiles}
          />
        ))}
      </FileList>
    </Container>
  );
};

export default NewFileUploadField;
