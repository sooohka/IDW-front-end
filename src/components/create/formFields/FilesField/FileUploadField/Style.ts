import styled from "styled-components";

const FileUploadField = styled.div`
  min-height: 150%;
`;

export { FileUploadField };
