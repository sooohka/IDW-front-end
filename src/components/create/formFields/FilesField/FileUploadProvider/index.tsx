import React, { createContext, useCallback, useContext, useReducer } from "react";

// actions
const DELETE_FILE = "file/deleteFile" as const;
const APPEND_FILES = "file/appendFiles" as const;
const UPDATE_FILE = "file/updateFile" as const;
const SET_ACCEPTING = "file/setIsAccepting" as const;

// action creators
const appendFiles = (files: TargetFile[]) => ({ type: APPEND_FILES, payload: { files } });
const updateFiles = (file: TargetFile) => ({ type: UPDATE_FILE, payload: { file } });
const deleteFile = (id: string) => ({ type: DELETE_FILE, payload: { id } });
const setIsAccepting = (isAccepting: boolean) => ({
  type: SET_ACCEPTING,
  payload: { isAccepting },
});

type Action =
  | ReturnType<typeof appendFiles>
  | ReturnType<typeof updateFiles>
  | ReturnType<typeof deleteFile>
  | ReturnType<typeof setIsAccepting>;

const initState: { files: TargetFile[]; isAccepting: boolean } = { files: [], isAccepting: false };
const reducer = (state = initState, action: Action): typeof initState => {
  switch (action.type) {
    case APPEND_FILES: {
      const { files } = action.payload;
      return { ...state, files: [...state.files, ...files] };
    }
    case UPDATE_FILE: {
      const { file } = action.payload;
      return { ...state, files: state.files.map((v) => (v.id === file.id ? file : v)) };
    }
    case DELETE_FILE: {
      const { id } = action.payload;
      return { ...state, files: state.files.filter((file) => file.id !== id) };
    }
    case SET_ACCEPTING: {
      const { isAccepting } = action.payload;
      return { ...state, isAccepting };
    }
    default:
      return state;
  }
};

interface FileUploadContextType {
  isAccepting: boolean;
  files: TargetFile[];
  handleFilesAppend: (files: TargetFile[]) => void;
  handleFileDelete: (id: string) => void;
  handleFileUpdate: (fils: TargetFile) => void;
  handleSetIsAccepting: (isAccepting: boolean) => void;
}

const FileUploadContext = createContext<FileUploadContextType>({
  isAccepting: false,
  files: [],
  handleFileDelete: () => {},
  handleFileUpdate: () => {},
  handleFilesAppend: () => {},
  handleSetIsAccepting: () => {},
});

const FileUploadProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initState);

  const handleFilesAppend = useCallback((files: TargetFile[]) => {
    dispatch(appendFiles(files));
  }, []);
  const handleFileDelete = useCallback((id: string) => {
    dispatch(deleteFile(id));
  }, []);
  const handleSetIsAccepting = useCallback((isAccepting: boolean) => {
    dispatch(setIsAccepting(isAccepting));
  }, []);
  const handleFileUpdate = useCallback((file: TargetFile) => {
    dispatch(updateFiles(file));
  }, []);

  const value = {
    files: state.files,
    isAccepting: state.isAccepting,
    handleFileDelete,
    handleFilesAppend,
    handleSetIsAccepting,
    handleFileUpdate,
  };
  return <FileUploadContext.Provider value={value}>{children}</FileUploadContext.Provider>;
};

const useFileUploadContext = () => useContext(FileUploadContext);

export { useFileUploadContext };
export default FileUploadProvider;
