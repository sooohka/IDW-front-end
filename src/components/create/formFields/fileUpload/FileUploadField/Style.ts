import styled from "styled-components";

const FileUploadField = styled.div`
  min-height: 150%;
`;

const DropZone = styled.div<{ isAccepting: boolean }>`
  border: 3px dashed;
  border-color: ${({ theme }) => theme.colors.primary};
  height: 15rem;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  flex-direction: column;
  & > * {
    transform: ${({ isAccepting }) => isAccepting && "scale(1.5)"};
    opacity: ${({ isAccepting }) => isAccepting && 0.3};
  }
  & > p {
    font-weight: bold;
    color: ${({ theme }) => theme.colors.primary};
  }
`;

export { FileUploadField, DropZone };
