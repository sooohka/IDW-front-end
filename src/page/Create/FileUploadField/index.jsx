import axios from "axios";
import { useField } from "formik";
import PropTypes from "prop-types";
import React, { useCallback, useEffect, useState } from "react";
import { useDropzone } from "react-dropzone";
import uuid from "react-uuid";
import useMount from "../../../utils/hooks/useMount";
import { readAsDataUrl } from "../../../utils/lib/file";
import Template from "./template";

const FileUploadField = ({ formikName, setIsFileUploading, buttonEl }) => {
  const [, , helpers] = useField(formikName);
  const [files, setFiles] = useState([]);
  // const [submittedFiles, setSubmittedFiles] = useState([]); // { name: "", url: "" } =>helper.setValue 해야함
  const [isAccepting, setIsAccepting] = useState(false);
  const { isMount } = useMount();
  const [isFolded, setIsFolded] = useState(true);

  const handleUpload = useCallback(async (_file, setProgress) => {
    const {
      file: { lastModified, lastModifiedDate, name, path, size, type },
      id: currentFileId,
      url: unused,
      isSubmitted,
      error: { status, message: m },
    } = _file;
    const copiedFile = { name, lastModified, lastModifiedDate, path, size, type };

    try {
      const fileDataUrl = await readAsDataUrl(_file.file);
      const request = { file: { name, contentType: type, dataUri: fileDataUrl } };

      if (process.env.REACT_APP_ENV !== "local") {
        alert("env=local입니다.");
        console.log("request:", request);
        return { hasError: false, message: "env=local", file: copiedFile, url: "" };
      }

      const response = await axios.post("https://dogemdas2c.execute-api.ap-northeast-2.amazonaws.com/v1", request, {
        onUploadProgress: (prog) => setProgress(Math.round(prog.loaded * 100) / prog.total),
      });

      const {
        message,
        result: { url, ContentType, bucketUrl, locations },
      } = response.data;

      const setFile = (v) => {
        if (v.id === currentFileId) return { ...v, isSubmitted: true, url, fullUrl: `${bucketUrl}${locations.small}` };
        return { ...v };
      };

      setFiles((prev) => prev.map(setFile));
      return {
        hasError: false,
        message: message || "submitted",
        file: { name, lastModified, lastModifiedDate, path, size, type },
        url: `${bucketUrl}${locations.small}`,
      };
    } catch (err) {
      console.error(err.message);
      // TODO: fileUploadWithProgress error handling
      const message = err.response?.data?.error?.message || err.message || "something went wrong😅 ";
      const setFile = (v) => {
        if (v.id === currentFileId) return { ...v, error: { status: true, message } };
        return { ...v };
      };
      setFiles(setFile);
      return { hasError: true, message, file: { name, lastModified, lastModifiedDate, path, size, type } };
    }
  }, []);

  useEffect(() => {
    if (buttonEl.current) buttonEl.current.scrollIntoView({ behavior: "smooth", block: "end" });
  }, [buttonEl, isFolded]);

  useEffect(() => {
    setIsFileUploading(true);
  }, [files, setIsFileUploading]);

  useEffect(() => {
    if (isMount) return;
    console.log(`%c state submittedFiles in FileUploadField changed`, "background-color:pink;font-size:15px;font-weight:bold;color:black");

    const isEveryFileSubmitted = files.every((v) => v.isSubmitted);
    if (isEveryFileSubmitted) setIsFileUploading(false);
    else helpers.setError("파일이 업로드중입니다.");

    const formattedFormikFile = files.map((v) => ({ name: v.name, url: v.url }));

    helpers.setValue(formattedFormikFile, true);
  }, [files, isMount, setIsFileUploading]);

  const onDrop = useCallback((acceptedFiles, rejectedFiles) => {
    // TODO: 에러메시지 개선
    if (rejectedFiles.length > 0) {
      console.log(rejectedFiles);
      const rejected = rejectedFiles.map((v) => ({ name: v.file.name, errors: v.errors[0].message }));
      alert(JSON.stringify({ rejectedFiles: rejected }, null, 2));
    }
    const formedAcceptedFiles = acceptedFiles.map((file) => ({ file, isSubmitted: false, id: uuid(), url: "", error: { status: false, message: "" } }));
    setFiles((prev) => [...prev, ...formedAcceptedFiles]);
    setIsAccepting(false);
  }, []);

  const onDragEnter = useCallback(() => setIsAccepting(true), []);
  const onDragLeave = useCallback(() => setIsAccepting(false), []);

  const handleValidation = useCallback(
    (_file) => {
      if (files.find(({ file }) => file.name === _file.name && file.lastModified === _file.lastModified)) return { message: "같은 파일은 업로드불가능합니다." };
      if (_file.size >= 5000000) return { message: "파일은 최대 4.9MB까지 업로드 가능합니다." };
      if (files.length > 50) return { message: "파일은 최대 50장 업로드 가능합니다." };

      return null;
    },
    [files]
  );

  const handleDelete = useCallback(
    (id) => () => {
      setFiles((prev) => prev.filter((file) => file.id !== id));
    },
    []
  );

  const { getRootProps, getInputProps } = useDropzone({
    accept: "image/*",
    maxFiles: 50,
    onDrop,
    onDragEnter,
    onDragLeave,
    validator: handleValidation,
  });

  return (
    <Template
      isFolded={isFolded}
      setIsFolded={setIsFolded}
      handleDelete={handleDelete}
      handleUpload={handleUpload}
      files={files}
      isAccepting={isAccepting}
      getRootProps={getRootProps}
      getInputProps={getInputProps}
    />
  );
};
FileUploadField.propTypes = {
  buttonEl: PropTypes.shape({ current: PropTypes.object }).isRequired,
  setIsFileUploading: PropTypes.func.isRequired,
  formikName: PropTypes.string.isRequired,
};
FileUploadField.defaultProps = {};

export default FileUploadField;
