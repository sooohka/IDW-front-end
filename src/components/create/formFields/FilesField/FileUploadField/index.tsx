import React, { useEffect } from "react";
import DropZone from "../DropZone";
import * as S from "../DropZone/Style";
import FileList from "../FileList";
import { useFileUploadContext } from "../FileUploadProvider";
import FileUploadWithProgress from "../FileUploadWIthProgress";

interface IProps {
  isFileUploading: boolean;
  setIsFileUploading: React.Dispatch<React.SetStateAction<boolean>>;
  handleFilesChange: (
    value: {
      url: string;
      name: string;
    }[],
  ) => void;
}
const FileUploadField: React.FC<IProps> = ({
  isFileUploading,
  setIsFileUploading,
  handleFilesChange,
}) => {
  const { files, handleFileDelete } = useFileUploadContext();

  useEffect(() => {
    if (isFileUploading) return;
    if (files.some((file) => !file.isSubmitted || file.errors.length > 0)) return;
    const formedFiles = files.map((imageFile) => ({
      name: imageFile.file.name,
      url: imageFile.url,
    }));
    handleFilesChange(formedFiles);
  }, [files, handleFilesChange, isFileUploading]);
  return (
    <S.FileUploadField>
      <DropZone />
      <FileList>
        {files.map((file) => (
          <FileUploadWithProgress
            setIsFileUploading={setIsFileUploading}
            imageFile={file}
            key={file.id}
          />
        ))}
      </FileList>
    </S.FileUploadField>
  );
};
export default FileUploadField;
