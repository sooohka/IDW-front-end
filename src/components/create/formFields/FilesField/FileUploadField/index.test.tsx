import FileUploadField from ".";
import { act, createWithThemeProvider, render } from "../../../../../__test__/test-utils";
import { FileUploadContext, FileUploadContextType } from "../FileUploadProvider";

describe("src/components/create/formFields/FilesField/FileUploadField", () => {
  let fileUploadFieldProps = {
    handleFilesChange: () => {},
    isFileUploading: false,
    setIsFileUploading: () => {},
  };

  let fileUploadContextValue: FileUploadContextType = {
    isAccepting: false,
    files: [],
    handleFileDelete: () => {},
    handleFileUpdate: () => {},
    handleFilesAppend: () => {},
    handleSetIsAccepting: () => {},
  };

  beforeEach(() => {
    fileUploadFieldProps = {
      handleFilesChange: jest.fn(),
      isFileUploading: false,
      setIsFileUploading: jest.fn(),
    };
    fileUploadContextValue = {
      files: [],
      isAccepting: false,
      handleFileDelete: jest.fn(),
      handleFileUpdate: jest.fn(),
      handleFilesAppend: jest.fn(),
      handleSetIsAccepting: jest.fn(),
    };
  });

  it("matches snapshot", () => {
    const component = createWithThemeProvider(<FileUploadField {...fileUploadFieldProps} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("doesn't update formik's files property when file is uploading to img server", () => {
    fileUploadFieldProps.isFileUploading = true;

    act(() => {
      render(
        <FileUploadContext.Provider value={fileUploadContextValue}>
          <FileUploadField {...fileUploadFieldProps} />
        </FileUploadContext.Provider>,
      );
    });

    expect(fileUploadFieldProps.handleFilesChange).not.toBeCalledTimes(1);
  });

  it("doesn't update formik's files property when file is not yet submitted ", () => {
    const file = new File(["blob"], "apple.png", { type: "image/png" });
    fileUploadContextValue.files = [{ errors: [], file, id: "1", isSubmitted: false, url: "" }];
    render(
      <FileUploadContext.Provider value={fileUploadContextValue}>
        <FileUploadField {...fileUploadFieldProps} />
      </FileUploadContext.Provider>,
    );
    expect(fileUploadFieldProps.handleFilesChange).not.toBeCalledTimes(1);
  });

  it("doesn't update formik's files property when some file has error", () => {
    const file = new File(["blob"], "apple.png", { type: "image/png" });
    fileUploadContextValue.files = [
      { errors: [{ code: "error", message: "error" }], file, id: "1", isSubmitted: true, url: "" },
    ];
    render(
      <FileUploadContext.Provider value={fileUploadContextValue}>
        <FileUploadField {...fileUploadFieldProps} />
      </FileUploadContext.Provider>,
    );
    expect(fileUploadFieldProps.handleFilesChange).not.toBeCalledTimes(1);
  });

  it("update formik's files property when single file that is submitted and has no error submitted", () => {
    const file = new File(["blob"], "apple.png", { type: "image/png" });
    fileUploadContextValue.files = [{ errors: [], file, id: "1", isSubmitted: true, url: "" }];
    render(
      <FileUploadContext.Provider value={fileUploadContextValue}>
        <FileUploadField {...fileUploadFieldProps} />
      </FileUploadContext.Provider>,
    );

    expect(fileUploadFieldProps.handleFilesChange).toBeCalledTimes(1);
  });

  it("update formik's files property when multiple files that is submitted and has no error submitted", () => {
    const file = new File(["blob"], "apple.png", { type: "image/png" });
    const file2 = new File(["blob"], "banana.png", { type: "image/png" });
    fileUploadContextValue.files = [
      { errors: [], file, id: "1", isSubmitted: true, url: "" },
      { errors: [], file: file2, id: "2", isSubmitted: true, url: "" },
    ];
    render(
      <FileUploadContext.Provider value={fileUploadContextValue}>
        <FileUploadField {...fileUploadFieldProps} />
      </FileUploadContext.Provider>,
    );

    expect(fileUploadFieldProps.handleFilesChange).toBeCalledWith(
      expect.arrayContaining([
        expect.objectContaining({ name: "apple.png" }),
        expect.objectContaining({ name: "banana.png" }),
      ]),
    );
    expect(fileUploadFieldProps.handleFilesChange).toBeCalledTimes(1);
  });

  it("gets multiple files that some are not submitted or has error", () => {
    const file = new File(["blob"], "apple.png", { type: "image/png" });
    const file2 = new File(["blob"], "banana.png", { type: "image/png" });
    const file3 = new File(["blob"], "carrot.png", { type: "image/png" });
    fileUploadContextValue.files = [
      { errors: [], file, id: "1", isSubmitted: false, url: "" },
      { errors: [], file: file2, id: "2", isSubmitted: true, url: "" },
      { errors: [], file: file3, id: "3", isSubmitted: true, url: "" },
    ];
    render(
      <FileUploadContext.Provider value={fileUploadContextValue}>
        <FileUploadField {...fileUploadFieldProps} />
      </FileUploadContext.Provider>,
    );

    expect(fileUploadFieldProps.handleFilesChange).not.toBeCalled();
  });
});
