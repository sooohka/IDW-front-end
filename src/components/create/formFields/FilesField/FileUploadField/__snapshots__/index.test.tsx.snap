// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/create/formFields/FilesField/FileUploadField matches snapshot 1`] = `
<div
  className="sc-bdvvtL jWuAIJ"
>
  <div
    className="sc-gsDKAQ bFxvOH"
    onBlur={[Function]}
    onClick={[Function]}
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    tabIndex={0}
  >
    <input
      accept="image/*"
      autoComplete="off"
      data-testid="dropzone-input"
      multiple={true}
      onChange={[Function]}
      onClick={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      tabIndex={-1}
      type="file"
    />
    <svg
      fill="#2d3436"
      height={50}
      width={50}
    >
      cloud-upload-alt-solid.svg
    </svg>
    <p>
      Drag and Drop or click here to upload
    </p>
  </div>
  <div
    className="sc-hKwDye gdeTzD"
  >
    <div
      className="sc-jRQBWg cgEMDs"
      data-testid="file-list-header"
      onClick={[Function]}
    >
      <h3
        className="sc-eCImPb fLbTcv"
      >
        파일들
      </h3>
      <div
        className="sc-iCfMLu cNAmWH"
      >
        <svg
          height={30}
          width={30}
        >
          sort-up-solid.svg
        </svg>
      </div>
    </div>
    <div
      className="sc-gKclnd cxPqzP"
      data-testid="file-list-content"
    >
      <div
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        <p
          className="sc-dkPtRN dsfKLW"
          fontSize="strongBody"
        >
          파일이 없습니다
        </p>
      </div>
    </div>
  </div>
</div>
`;
